src/odemis/driver/smaract.py

```diff
@@ -4913,6 +4922,406 @@ def _cancelReference(self, future):
+            IOError: if referencing failed due to hardware
+            CancelledError if was cancelled
+        """
+        try:
+            # TODO: can we leave the position polling thread on? It does not seem
+            # to interfere with the referencing procedure.
+            if self._polling_thread:
+                self._polling_thread.cancel()
+                self._polling_thread = None
+
+            with future._moving_lock:
+                if future._must_stop.is_set():
+                    raise CancelledError()
+                # Reset reference so that if it fails, it states the axes are not referenced (anymore)
+                self.referenced._value = {a: False for a in self._channels.keys()}
+                # Cannot go immediately to automatic adjustment --> first switch to manual adjustment
+                si.SetProperty_i32(self.id, si.Property.EVENT_NOTIFICATION_ENABLED << 16 | ps.EventType.AF_ADJUSTMENT_PROGRESS, si.ENABLED)
+                # Enable channels
+                for channel_index in self._channels.values():
+                    si.SetProperty_i32(self.id, si.EPK(ps.Property.CH_ENABLED, channel_index, 0), si.ENABLED)
+                si.SetProperty_i32(self.id, si.EPK(ps.Property.AF_ADJUSTMENT_STATE, 0, 0), ps.AdjustmentState.MANUAL_ADJUST)
+            self._wait_for_progress_event(ps.AdjustmentState.MANUAL_ADJUST, timeout=600)
+            logging.debug("Manual adjustment finished.")
+
+            with future._moving_lock:
+                if future._must_stop.is_set():
+                    raise CancelledError()
+                # Activate working distance
+                si.SetProperty_i32(self.id, si.EPK(ps.Property.SYS_WORKING_DISTANCE_ACTIVATE, 0, 0), ps.WorkingDistanceShrinkMode.LEFT_RIGHT)
+            # attribute is write-only and there is no corresponding event type (yet some waiting time is necessary)
+            # --> wait 1 s for command to be processed
+            if future._must_stop.wait(1):
+                raise CancelledError()
+            logging.debug("Working distance active.")
+
+            with future._moving_lock:
+                if future._must_stop.is_set():
+                    raise CancelledError()
+                # Switch to automatic adjustment
+                si.SetProperty_i32(self.id, si.EPK(ps.Property.AF_ADJUSTMENT_STATE, 0, 0), ps.AdjustmentState.AUTO_ADJUST)
+            self._wait_for_progress_event(ps.AdjustmentState.DISABLED, timeout=600)  # state will be DISABLED when done
+            logging.debug("Finished referencing.")
+
+            # We could save the referencing parameters to memory here. This could be useful for the validation,
+            # because it initializes the system by loading the stored referencing parameters. However, we don't
+            # currently use validation, so there is no need to store the values every time we reference.
+        except si.Error as error:
+            logging.debug(self._getReadableSIError(error))
```

Comment:
Drop. Maybe, if that's not in `str(error)`, also include the error name in line 5145.